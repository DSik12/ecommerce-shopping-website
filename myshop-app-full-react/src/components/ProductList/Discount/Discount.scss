@import "../../shared/ScssBreakpoint/Breakpoint.scss";
* {
    margin: 0;
    padding: 0;
    box-sizing: border-box;
}

.formcheck {
    text-align: left;
}

.checklist {
    h3 {
        padding: 30px 0 5px 10px;
    }
}

@media (min-width: $tablet-portrait) {
    .inFilter {
        display: flex;
        width: 100%;
        input {
            margin: 10px;
        }
        label {
            margin: 5px 5px 10px 10px;
        }
    }
}
