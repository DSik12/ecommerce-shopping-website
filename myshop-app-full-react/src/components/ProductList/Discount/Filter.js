import React from 'react';

const Filter = (props) => {
    return (
        <span className="inFilter">
            <input type="checkbox" name="check" className='disable' id={props.id} />
            <label htmlFor={props.id}>{props.discountLabel}</label>
        </span>
    )
}

export default Filter;