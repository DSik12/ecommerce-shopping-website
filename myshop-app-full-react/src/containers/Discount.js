/* eslint-disable default-case */
import React, { Component } from 'react'
import Filter from '../components/ProductList/Discount/Filter';

export default class Discount extends Component {

    constructor() {
        super();
        this.apiURL = "https://ecom-app-rest-api-v1.herokuapp.com/products";
        this.lowToHigh = "lowToHigh";
        this.highToLow = "highToLow";
        this.bestSeller = "bestSeller";
        this.featured = "featured";
        this.productApi = null;
        this.checkboxes = [
            { id: "Discount1", discountLabel: "10% or more" },
            { id: "Discount2", discountLabel: "20% or more" },
            { id: "Discount3", discountLabel: "30% or more" },
            { id: "Discount4", discountLabel: "40% or more" },
            { id: "Discount5", discountLabel: "Below 10%" }
        ];
    }

    checkOnlyOne = (checkbox) => {

        const checkboxes = document.getElementsByName("check");
        let flag = null;
        let getProduct = null;
        checkboxes.forEach((item) => {
            if (item !== checkbox) {
                item.checked = false;
            } else {
                getProduct = item;
                flag = item.checked;
            }
        });

        const checkedID = getProduct.id;


        const sortAndChecked = (lowerLimit, upperLimit) => {

            const search = window.location.search;
            if (search) {
                const existingQuery = search.substring(1, search.length);
                if (!existingQuery.includes("discount")) {
                    const productApi = `${existingQuery}&discountApplicable_gte=${lowerLimit}&discountApplicable_lte=${upperLimit}`;
                    this.props.navigate("/products", productApi);
                }
                else {
                    const productApi = existingQuery.replace(/discountApplicable_gte=[0-9]*&discountApplicable_lte=[0-9]*/, `discountApplicable_gte=${lowerLimit}&discountApplicable_lte=${upperLimit}`);
                    this.props.navigate("/products", productApi);
                }
            }
            else {
                const productApi = `discountApplicable_gte=${lowerLimit}&discountApplicable_lte=${upperLimit}`;
                this.props.navigate("/products", productApi);
            }
        };

        if (flag === true) {
            switch (checkedID) {
                case "Discount1":
                    return sortAndChecked(10, 100);
                case "Discount2":
                    return sortAndChecked(20, 100);
                case "Discount3":
                    return sortAndChecked(30, 100);
                case "Discount4":
                    return sortAndChecked(40, 100);
                case "Discount5":
                    return sortAndChecked(0, 10);
            }
        }
        else {
            const search = window.location.search;
            const existingQuery = search.substring(1, search.length);

            if(existingQuery.match(/^discountApplicable_gte=[0-9]*&discountApplicable_lte=[0-9]*/)){
                if(existingQuery.match(/discountApplicable_gte=[0-9]*&discountApplicable_lte=[0-9]*&/)){
                    const productApi = existingQuery.replace(/discountApplicable_gte=[0-9]*&discountApplicable_lte=[0-9]*&/, "");
                    this.props.navigate("/products", productApi);
                }
                else{
                    const productApi = existingQuery.replace(/discountApplicable_gte=[0-9]*&discountApplicable_lte=[0-9]*/, "");
                    this.props.navigate("/products", productApi);
                }
                
            }
            else if(existingQuery.match(/discountApplicable_gte=[0-9]*&discountApplicable_lte=[0-9]*$/)){
                const productApi = existingQuery.replace(/&discountApplicable_gte=[0-9]*&discountApplicable_lte=[0-9]*/, "");
                this.props.navigate("/products", productApi);
            }
            else{
                const productApi = existingQuery.replace(/&discountApplicable_gte=[0-9]*&discountApplicable_lte=[0-9]*/, "");
                this.props.navigate("/products", productApi);
            }
        }
    };

    handleDiscount(event) {
        event.stopPropagation();
        this.checkOnlyOne(event.target);
    }

    render() {

        const checkboxes = this.checkboxes.map((checkbox) => {
            return <Filter id={checkbox.id} discountLabel={checkbox.discountLabel} key={checkbox.id} />
        })
        return (
            <section className="checklist">
                <h3>Filter</h3>
                <h3>DISCOUNT</h3>

                <form className="formcheck" id="discountForm" onChange={this.handleDiscount.bind(this)} data-testid="discountEl">
                    {checkboxes}

                </form>
            </section>

        )
    }
}

