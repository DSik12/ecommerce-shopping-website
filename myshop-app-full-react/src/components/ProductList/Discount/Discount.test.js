import React from 'react';
import ReactDom from 'react-dom';
import { BrowserRouter } from 'react-router-dom';
import {render, screen} from '@testing-library/react';
import renderer from 'react-test-renderer';
import Discount from '../../../containers/Discount';

describe("Discount Component", () => {
    it('renders the component properly', () => {
        const sectionEl = document.createElement('section');
        ReactDom.render(<BrowserRouter><Discount/></BrowserRouter>, sectionEl);
    });

    it('should have Filter as text', () => {
        render(<BrowserRouter><Discount/></BrowserRouter>);
        const filterText = screen.getByText('Filter');
        expect(filterText).toBeInTheDocument();
    });

    it('should have DISCOUNT as text', () => {
        render(<BrowserRouter><Discount/></BrowserRouter>);
        const discountText = screen.getByText('DISCOUNT');
        expect(discountText).toBeInTheDocument();
    });

    it('should display discount filters correctly', () => {
        document.body.innerHTML = `
        <span class="inFilter">
            <input type="checkbox" name="check" class='disable' id="Discount1" />
            <label htmlFor="Discount1">10% or more</label>
        </span>
        <span class="inFilter">
            <input type="checkbox" name="check" class='disable' id="Discount2" />
            <label htmlFor="Discount2">20% or more</label>
        </span>
        `;
        require('../../../containers/Discount.js');
    });

    it('should have right Discount component snapshot', () => {
        const tree = renderer.create(<BrowserRouter><Discount/></BrowserRouter>).toJSON();
        expect(tree).toMatchSnapshot();
    });
    
});
