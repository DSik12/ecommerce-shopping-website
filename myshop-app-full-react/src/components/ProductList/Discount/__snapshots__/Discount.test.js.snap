// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Discount Component should have right Discount component snapshot 1`] = `
<section
  className="checklist"
>
  <h3>
    Filter
  </h3>
  <h3>
    DISCOUNT
  </h3>
  <form
    className="formcheck"
    data-testid="discountEl"
    id="discountForm"
    onChange={[Function]}
  >
    <span
      className="inFilter"
    >
      <input
        className="disable"
        id="Discount1"
        name="check"
        type="checkbox"
      />
      <label
        htmlFor="Discount1"
      >
        10% or more
      </label>
    </span>
    <span
      className="inFilter"
    >
      <input
        className="disable"
        id="Discount2"
        name="check"
        type="checkbox"
      />
      <label
        htmlFor="Discount2"
      >
        20% or more
      </label>
    </span>
    <span
      className="inFilter"
    >
      <input
        className="disable"
        id="Discount3"
        name="check"
        type="checkbox"
      />
      <label
        htmlFor="Discount3"
      >
        30% or more
      </label>
    </span>
    <span
      className="inFilter"
    >
      <input
        className="disable"
        id="Discount4"
        name="check"
        type="checkbox"
      />
      <label
        htmlFor="Discount4"
      >
        40% or more
      </label>
    </span>
    <span
      className="inFilter"
    >
      <input
        className="disable"
        id="Discount5"
        name="check"
        type="checkbox"
      />
      <label
        htmlFor="Discount5"
      >
        Below 10%
      </label>
    </span>
  </form>
</section>
`;
